### PR TITLE
feat(model): add MiniCPM-o 4.5 omni model support (3-stage pipeline + async_chunk streaming)

### PR DESCRIPTION
**Pull request by luke-n-alpha**
## Summary

Adds full MiniCPM-o 4.5 support to vllm-omni following the established 3-stage pipeline pattern.

Related: #1182 (parallel effort by @Allyyi — we have been coordinating; see that issue for context)

### What's included

**Core model** (`vllm_omni/model_executor/models/minicpm_o/`)
- **Thinker**: Idefics2 vision encoder + Whisper audio encoder + Qwen3 LLM backbone
- **Talker**: MiniCPMTTS AR codec generator (`num_vq=1`, single RVQ layer vs Qwen3's 8)
- **Code2Wav**: CosyVoice2 flow matching + HiFi-GAN vocoder

**Stage configs** (`vllm_omni/model_executor/stage_configs/`)
| Config | Use case | GPU |
|--------|---------|-----|
| `minicpmo.yaml` | Single 24 GB GPU | RTX 3090 ×1 |
| `minicpmo_48gb_2gpu.yaml` | 2× 24 GB GPU, sync | RTX 3090 ×2 |
| `minicpmo_async_chunk.yaml` | 2× 24 GB GPU, streaming | RTX 3090 ×2 |

**Stage input processor** (`vllm_omni/model_executor/stage_input_processors/minicpm_o.py`)
- `thinker2talker_sync` / `talker2code2wav` (synchronous path)
- `thinker2talker_async_chunk` / `talker2code2wav_async_chunk` (streaming path)
- `_find_tts_bound()`: locates `<|tts_bos|>`…`<|tts_eos|>` boundary in Thinker output — required because MiniCPM-o embeds TTS tokens inline in the generation sequence (unlike Qwen3-Omni which uses the full sequence)
- `_ensure_list()`: handles `ConstantList` / tensor codec token normalization (`num_vq=1` means all codec tokens are 1D lists)

**Serving** (`vllm_omni/entrypoints/openai/serving_chat.py`)
- Auto-inject `use_tts_template=True` for MiniCPM-o requests via chat template kwargs

**Tests**
- `tests/e2e/offline_inference/test_minicpm_o.py` — smoke tests (L1/L2/L3)
- `tests/e2e/stage_configs/minicpm_o_ci.yaml` — CI config with `load_format: dummy`
- `tests/model_executor/models/minicpm_o/` — unit tests (config, components, stage processors)

**Examples**
- `examples/offline_inference/minicpm_o/` — `end2end.py` (sync `Omni`), `end2end_async_chunk.py` (async `AsyncOmni`), convenience shell scripts
- `examples/online_serving/minicpm_o/` — README, BENCHMARK, conversation/language/VoiceBench scripts

### Key architectural differences from Qwen3-Omni

| | MiniCPM-o 4.5 | Qwen3-Omni |
|--|---------------|-----------|
| num_vq (RVQ layers) | **1** | 8 |
| Hidden state key | `thinker_hidden_states` | `thinker_prefill_embeddings` + `thinker_decode_embeddings` |
| Codec token shape | 1D flat | 8×N matrix |
| TTS boundary | `_find_tts_bound()` required | no (full sequence) |
| async_chunk | ✅ this PR | ✅ upstream |
| Config class | custom (not in transformers) | transformers |

### Benchmark results (2× RTX 3090, async_chunk mode)

| Language | STT Accuracy | Avg Latency | TTFP |
|----------|:------------|:-----------|:-----|
| English | 92% word accuracy | 2.3s | ~0.07s |
| Chinese | 95% semantic similarity | 1.5s | ~0.07s |
| Korean | text OK / TTS failed* | — | — |

VoiceBench (EN, rule-based scoring): 98.0% avg, 98.7% pass rate

*Korean TTS fails because CosyVoice2 was not trained on Korean. Known limitation, tracked separately.

### Known limitations

- **Stop token 6561** rarely generated by Talker → audio length fixed at ~20s. Root cause: model conditioning issue, not a code bug. `_trim_silence()` post-processing applied.
- **Korean TTS**: CosyVoice2 not trained on Korean → requires fine-tuning.
- **NVLink**: `NCCL_P2P_DISABLE=1` required for RTX 3090 without NVLink.

### Incidental bugfixes (try/except StopIteration → next(..., None))

While implementing, found the same `StopIteration` pattern in existing models. Cleaned up:
- `models/mammoth_moda2/mammoth_moda2.py`
- `models/mimo_audio/mimo_audio.py`
- `models/qwen2_5_omni/qwen2_5_omni.py`
- `models/qwen3_tts/qwen3_tts_code2wav.py`, `qwen3_tts_tokenizer.py`

### Testing

Validated on local 2× RTX 3090 (no NVLink):
- [x] Single GPU (24 GB) — sync inference
- [x] Dual GPU (48 GB total) — sync inference
- [x] Dual GPU — async_chunk streaming (TTFP ~0.07s)
- [x] Text + image + audio input modalities
- [x] Text-only output (`--modalities text`)
- [x] Multi-prompt batch (`py_generator` mode)
- [x] Unit tests (config, components, stage processors)

### Code review

8-pass adversarial headless review performed (accuracy, completeness, consistency, pattern compliance, upstream validation, operational completeness, test validity, comprehensive re-check). 2 consecutive clean passes achieved.

---

> **Note on #1182**: @Allyyi has been working on MiniCPM-o in parallel. We have been sharing findings in that issue. If both PRs are submitted, we are happy to coordinate on deduplication or merge the best parts of each implementation.

### CODE REVIEW

File: vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py [hunk 1]
```diff
@@ -773,10 +773,8 @@ def get_input_embeddings(self, input_ids: torch.Tensor, multimodal_embeddings=No
         # to populate `inputs_embeds` buffers, so we provide a dummy tensor.
         if self.model_stage == "dit":
             hidden_size = int(self.vllm_config.model_config.get_hidden_size())
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py:775]: this file looks irrelavent to the goal of this PR

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code2wav.py [hunk 2]
```diff
@@ -45,12 +45,17 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code2wav.py:45]: this file looks irrelavent to the goal of this PR

File: vllm_omni/model_executor/models/qwen2_5_omni/qwen2_5_omni.py [hunk 3]
```diff
@@ -151,13 +151,17 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
     # -------------------- Device utilities --------------------
     @staticmethod
     def _module_device(module: nn.Module) -> torch.device:
-        try:
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen2_5_omni/qwen2_5_omni.py:154]: this file looks irrelavent to the goal of this PR

**Comment by hsliuustc0106** [comment]
fix all of the basic conflicts first

**Comment by luke-n-alpha** [comment]
Closing — submitted prematurely. Will reopen when ready for review.

**Comment by luke-n-alpha** [comment]
Apologies for the noise — this PR was submitted unintentionally before final review. Closing for now and will reopen when properly ready.
